### PR TITLE
Consolidate version-numbering semantics

### PR DESCRIPTION
**Pull request by DaveCTurner**
At the moment we interpret version numbers, and our expectations for things like which unreleased versions there are and on which branches their code lives, in a number of different places. These various interpretations are not obviously equivalent, and the bump to 6.0.1 certainly uncovered some issues in this code. The way that some of the issues were fixed will bite us at the bump to 6.0.2.

The plan is to consolidate this logic into one place, and test it more thoroughly, to avoid this kind of thing catching us out in future.

### CODE REVIEW

File: build.gradle [hunk 1]
```diff
@@ -252,30 +203,25 @@ subprojects {
...
   }
+  final def bwcReleaseSnapshot = versionCollection.BWCSnapshotForCurrentMajor
+  if (bwcReleaseSnapshot != null) {
+    ext.projectSubstitutions["org.elasticsearch.distribution.deb:elasticsearch:${bwcReleaseSnapshot}"] = ':distribution:bwc-release-snapshot'
+    ext.projectSubstitutions["org.elasticsearch.distribution.rpm:elasticsearch:${bwcReleaseSnapshot}"] = ':distribution:bwc-release-snapshot'
+    ext.projectSubstitutions["org.elasticsearch.distribution.zip:elasticsearch:${bwcReleaseSnapshot}"] = ':distribution:bwc-release-snapshot'
+  }
+  final def bwcPreviousReleaseSnapshot = versionCollection.BWCSnapshotForPreviousMinorOfCurrentMajor
+  if (bwcPreviousReleaseSnapshot != null) {
+    throw new NotImplementedException(
```
Comment (nik9000) [inline build.gradle]: :+1:

File: buildSrc/src/main/groovy/org/elasticsearch/gradle/vagrant/VagrantTestPlugin.groovy [hunk 2]
```diff
@@ -107,7 +107,8 @@ class VagrantTestPlugin implements Plugin<Project> {
         if (upgradeFromVersion == null) {
             String firstPartOfSeed = project.rootProject.testSeed.tokenize(':').get(0)
             final long seed = Long.parseUnsignedLong(firstPartOfSeed, 16)
-            upgradeFromVersion = project.indexCompatVersions[new Random(seed).nextInt(project.indexCompatVersions.size())]
+            final def indexCompatVersions = project.versionCollection.versionsIndexCompatibleWithCurrent
```
Comment (nik9000) [inline buildSrc/src/main/groovy/org/elasticsearch/gradle/vagrant/VagrantTestPlugin.groovy]: We've been mostly avoiding `def` in `.groovy` files so they are more "java-like". Can you use the type here?

**Review by nik9000 — APPROVED** [review]

File: build.gradle (hunk 1 above)
Comment (DaveCTurner) [reply build.gradle]: Gah, the version bump to 6.2.0 just got pushed, so now I can't merge this without throwing this exception. :(

File: buildSrc/src/main/groovy/org/elasticsearch/gradle/vagrant/VagrantTestPlugin.groovy (hunk 2 above)
Comment (DaveCTurner) [reply buildSrc/src/main/groovy/org/elasticsearch/gradle/vagrant/VagrantTestPlugin.groovy]: Sure, no probs

**Comment by DaveCTurner** [comment]
@martijnvg Jenkins and I are happy about this. Hopefully your test run is too.

**Review by martijnvg — APPROVED** [review]
@DaveCTurner My test run (`gradle bwcTest`) was happy too.

**Comment by DaveCTurner** [comment]
Merged and backported:

89ba8996c6c735fa6fa0bca1a5cf4268b5daf07e on `master`
27ae6acceb1a9e3ecc2c57e3c7fc92efc04112c9 on `6.x`
a762da6880716e53bcb39bf7aad00456d935d4ef on `6.1`
2c7268133c5f918d48d6b6f1a255dc4657c8b7c9 on `6.0`

**Pull request closed without merging**